port/unix/omrsysinfo.c

```diff
@@ -2799,24 +2800,50 @@ omrsysinfo_get_number_CPUs_by_type(struct OMRPortLibrary *portLibrary, uintptr_t
 			int32_t rc = 0;
 			int64_t cpuQuota = 0;
 			uint64_t cpuPeriod = 0;
-			int32_t numItemsToRead = 1; /* cpu.cfs_quota_us and cpu.cfs_period_us files each contain only one integer value */
 
-			/* If either file read fails, ignore cgroup cpu quota limits and continue */
-			rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_CPU, "cpu.cfs_quota_us", numItemsToRead, "%ld", &cpuQuota);
-			if (0 == rc) {
-				rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_CPU, "cpu.cfs_period_us", numItemsToRead, "%lu", &cpuPeriod);
+			if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V1_AVAILABLE)) {
+				/* cpu.cfs_quota_us and cpu.cfs_period_us files each contain only one integer value. */
+				int32_t numItemsToRead = 1;
+
+				/* If either file read fails, ignore cgroup cpu quota limits and continue. */
+				rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_CPU, "cpu.cfs_quota_us", numItemsToRead, "%ld", &cpuQuota);
 				if (0 == rc) {
-					int32_t numCpusQuota = (int32_t) (((double) cpuQuota / cpuPeriod) + 0.5);
+					rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_CPU, "cpu.cfs_period_us", numItemsToRead, "%lu", &cpuPeriod);
+				}
+			} else {
+				/* Cgroup v2 is available. Read cpu.max file which contains the quota and period
+				 * in the format "$QUOTA $PERIOD". The value "max" for $QUOTA indicates no limit.
+				 */
+				int32_t numItemsToRead = 2;
+				char quotaString[MAX_64BIT_INT_LENGTH];
+				uint64_t quotaVal = 0;
 
-					if ((cpuQuota > 0) && (numCpusQuota < toReturn)) {
-						toReturn = numCpusQuota;
-						/* If the CPU quota rounds down to 0, then just return 1 as the closest usable value */
-						if (0 == toReturn) {
-							toReturn = 1;
+				rc = readCgroupSubsystemFile(portLibrary, OMR_CGROUP_SUBSYSTEM_CPU, "cpu.max", numItemsToRead, "%s %lu", &quotaString, &cpuPeriod);
+				if (0 != rc) {
+					Trc_PRT_sysinfo_get_number_CPUs_by_type_read_failed("cpu.max", rc);
+				} else {
+					rc = scanCgroupIntOrMax(portLibrary, quotaString, &quotaVal);
+					if (0 == rc) {
+						if (UINT64_MAX == quotaVal) {
+							cpuQuota = -1;
+						} else {
+							cpuQuota = (int64_t)quotaVal;
 						}
 					}
 				}
 			}
+
+			if (0 == rc) {
+				int32_t numCpusQuota = (int32_t) (((double) cpuQuota / cpuPeriod) + 0.5);
```

Comment:
Extra spaces are not needed. `cast` trumps `division (/)`. Our intent is to cast the result of `(cpuQuota / cpuPeriod)`. Extra brackets will enforce that. Otherwise, `cast` will first apply to `cpuQuota` and then the division will occur.
```suggestion
				int32_t numCpusQuota = (int32_t)(((double)(cpuQuota / cpuPeriod)) + 0.5);
```

Reply:
What is the significance of `0.5` here?

Reply:
Isn't the cast there to ensure floating point division, I assume for more precision? So a cpuQuota of 1500 and cpuPeriod of 1000 will not result in 1 cpu, but 1.5 instead. Also, the + 0.5 is for rounding the double before integer truncation, so the 1.5 + 0.5 = 2 instead of 1

Reply:
You're correct. Both `cpuQuota` and `cpuPeriod` are integers. Cast happens before division. Converting an integer to double before division should lead to a floating point division instead of integer division. You are also correct about +0.5 being used for rounding-off. Ignore my request for `extra brackets`; it will lead to incorrect behaviour.